checks.d/snmp.py

```diff
@@ -123,49 +139,89 @@ def check_table(self, instance, oids, lookup_names):
         dict[oid/metric_name][row index] = value
         In case of scalar objects, the row index is just 0
         '''
+        # UPDATE: We used to perform only a snmpgetnext command to fetch metric values.
+        # It returns the wrong value when the OID passeed is referring to a specific leaf.
+        # For example:
+        # snmpgetnext -v2c -c public localhost:11111 1.36.1.2.1.25.4.2.1.7.222
+        # iso.3.6.1.2.1.25.4.2.1.7.224 = INTEGER: 2
+        # SOLUTION: perform a snmget command and fallback with snmpgetnext if not found
         transport_target = self.get_transport_target(instance, self.TIMEOUT, self.RETRIES)
         auth_data = self.get_auth_data(instance)
 
-        snmp_command = self.cmd_generator.nextCmd
-
-    	first_oid = 0
-    	all_binds = []
+        first_oid = 0
+        all_binds = []
         results = defaultdict(dict)
 
-    	while first_oid < len(oids):
-	        error_indication, error_status, error_index, var_binds = snmp_command(
-	            auth_data,
-	            transport_target,
-	            *(oids[first_oid:first_oid+OID_BATCH_SIZE]),
-	            lookupValues = lookup_names,
-	            lookupNames = lookup_names
-	            )
-                if error_indication:
-                    message = "{0} for instance {1}".format(error_indication,
-                                                          instance["ip_address"])
-                    instance["service_check_error"] = message
-                    raise Exception(message)
-                else:
-                    if error_status:
-                        message = "{0} for instance {1}".format(error_status.prettyPrint(),
-                                                                       instance["ip_address"])
-                        instance["service_check_error"] = message
-                        self.log.warning(message)
-                    else:
-                        all_binds.extend(var_binds)
-    		first_oid = first_oid + OID_BATCH_SIZE
-
-        for table_row in all_binds:
-            for result_oid, value in table_row:
-                if lookup_names:
-                    object = result_oid.getMibSymbol()
-                    metric =  object[1]
-                    indexes = object[2]
-                    results[metric][indexes] = value
+        while first_oid < len(oids):
+
+            # Start with snmpget command
+            snmp_command = self.cmd_generator.getCmd
+            error_indication, error_status, error_index, var_binds = snmp_command(
+                auth_data,
+                transport_target,
+                *(oids[first_oid:first_oid + OID_BATCH_SIZE]),
+                lookupValues=lookup_names,
+                lookupNames=lookup_names)
+
+            # Raise on error_indication
+            self.raise_on_error_indication(error_indication, instance)
+
+            # Continue on error_status
+            if error_status:
```

Comment:
`report_error_indication` is raising an Exception and stops the check execution.

The reason why I didn't factorize `report_error_status` is because it has a `continue` (c.f your comment below) statement which refers directly to the upper level `for` loop.